### ISSUE TITLE
Bayesian Neural Network doesn't work for my dataset

### ISSUE THREAD

**Issue by Nikolin4** [pyro-ppl/numpyro#774]
Hi there!

I am writting to you because  bayesian neural model does not produce the desired output.
I have one hot encoding data, also X.shape = (N,20), also the data has 20 features, each for an amino acid.
The targets are  Y.shape = (N,2), also one feature stands for phi angles, and the other feature stands for psi angles.
N stands for the numbers of the rows.

My objectives are to predict the phi, psi, kappa, and mu where mu and kappa are the parameters for von Misses distribution,
phi and psi are the torsion angles for proteins.  (kappa.shape, mu.shape = (N, 2)## phi.shape,  psi.shape = (N,1))

Can anyone help develop this model using latent variables?
I cannot share the code from here, but let me know how we can work on it

This model is taken from http://pyro.ai/numpyro/examples/bnn.html .
The model in the following is written directly here without having tested. 
(I haven't shared the code, but this is what I use)
Please contribute to build this model, maybe by adding latent variables, or anything might be useful.
I can't get my head around this problem
Hope to hear from you

```python
def model(X, Y, D_H):
	D_X, D_Y = X.shape[1], 2
  w1 = numpyro.sample("w1", dist.Normal( 0,1 ), sample_shape (D_X, D_H))
  b1 = numpyro.sample("b1", dist.Normal(0,1), sample_shape(D_H,))
  z1 = softmax( jnp.matmul(X,w1) +b1)

  w2 = numpyro.sample("w2", dist.Normal( 0,1 ), sample_shape (D_H D_H))
  b2 = numpyro.sample("b2", dist.Normal(0,1), sample_shape(D_H,))
  z2 = softmax( jnp.matmul(z2,w2) +b2)


  w3 = numpyro.sample("w3", dist.Normal(0,1   ), sample_shape=(D_H, 2*D_Y)  )
  b3 = numpyro.sample("b3", dist.Normal(0,1), sample_shape = (2*D_Y, ) )
  z3 = jnp.matmul(z2, w3) + b3
  kappa = numpyro.deterministic("kappa", sigmoid(z3[ : ,  :2 ] ) ) #Activation on kappas needed
  mu = numpyro.deterministic("mu",  z3[ : , 2: ] ) #No activation needed
  numpyro.sample('phi', dist.VonMises(mus[: ,0], kappas[: , 0], obs= Y[: ,0]) 
  numpyro.sample('psi", dist.VonMises(mus[:, 1], kappas[:, 1], obs= Y[: ,1])
```

**Comment by martinjankowiak** [comment]
neural network architectures/assumptions need to be tailored to the data. the neural network in the example is very simple and is tailored to the 1 dimensional input data that is used. one needs to take care, for example, that each layer has bias units. there are different ways to do this, including encoding bias via a all-1s feature. one should also be aware that there's no reason to expect HMC to work out of the box if you make the dimensions of the problem (e.g. number of data points or number of hidden units) very large. without more details about your specific problem it's hard to give any additional advice.

**Comment by Nikolin4** [comment]
> neural network architectures/assumptions need to be tailored to the data. the neural network in the example is very simple and is tailored to the 1 dimensional input data that is used. one needs to take care, for example, that each layer has bias units. there are different ways to do this, including encoding bias via a all-1s feature. one should also be aware that there's no reason to expect HMC to work out of the box if you make the dimensions of the problem (e.g. number of data points or number of hidden units) very large. without more details about your specific problem it's hard to give any additional advice.

I would like to give more details, but I can't share everything online.   Could you think of one possibility. Data must be 20 dimentional because of amino acids.

**Comment by martinjankowiak** [comment]
how many datapoints do you have? what are DX DY DH? why are you using a softmax non-linearity? that seems strange. have you tried fewer layers? have you tried using a subset of your data? what do your observation model look like?

**Comment by Nikolin4** [comment]
> how many datapoints do you have? what are DX DY DH? why are you using a softmax non-linearity? that seems strange. have you tried fewer layers? have you tried using a subset of your data? what do your observation model look like?

Input data is  X, and it is one hot encoding.
It has a shape (50 000, 20)

Out target data (50 000, 2) Representing the torsion angles.
Data split into train and test: x_train = X[ :25000, :] the rest is the X_test.
Y_train = Y[: 25000, :], the rest for the Y_test.
Y_train.shape = (25000, 2), same for Y_test.
X_test.shape =(25000, 20).
Also from model, D_X =20, D_Y =2, D_H are the Number of nodes in each layer.

I have used small subsets of data, but The model doesn’t predict what is desired.

I think that must be developed because the model is too simple for the data.

The output are so bad,I could show them, but how since don’t want from online

**Comment by martinjankowiak** [comment]
well apart from other possible issues with your model the large number of datapoints is probably a significant problem. the loss surface you get with N large is going to be very sharp. as a consequence you're probably seeing the HMC step size get very small, e.g. 10^-5 or 10^-7. as a consequence HMC probably does a very poor job of exploring the posterior. unfortunately there are no easily solutions here. bayesian neural networks are very much an active research field and, generally speaking, they do not work out of the box (if at all).

**Comment by fehiepsi** [comment]
How about using SVI instead of MCMC? You can mimic the example in [random_flax_module](http://num.pyro.ai/en/stable/primitives.html#random-flax-module): using Flax to construct a neural network that predicts `mu` and `kappa`, then set `Normal()` priors for weights/biases. My main concern is if `phi` and `psi` are latent variables, it will be tricky to run MCMC/SVI unless there is a way to reparameterize `kappas` (you can reparameterize `mus` by using the fact that `VonMises(mus, kappas) = mus + VonMises(0, kappas)`). In this case, I think you will need to construct a VonMises guide, use Uniform prior (i.e. `kappas = 0` in model), and request for a KL divergence implementation of Uniform and VonMises (along the line of [this reference](https://arxiv.org/abs/1804.00891) e.g.). If `phi` and `psi` are observed, then no need to worry about that.

**Comment by Nikolin4** [comment]
> How about using SVI instead of MCMC? You can mimic the example in [random_flax_module](http://num.pyro.ai/en/stable/primitives.html#random-flax-module): using Flax to construct a neural network that predicts `mu` and `kappa`, then set `Normal()` priors for weights/biases. My main concern is if `phi` and `psi` are latent variables, it will be tricky to run MCMC/SVI unless there is a way to reparameterize `kappas` (you can reparameterize `mus` by using the fact that `VonMises(mus, kappas) = mus + VonMises(0, kappas)`). In this case, I think you will need to construct a VonMises guide, use Uniform prior (i.e. `kappas = 0` in model), and request for a KL divergence implementation of Uniform and VonMises (along the line of [this reference](https://arxiv.org/abs/1804.00891) e.g.). If `phi` and `psi` are observed, then no need to worry about that.

I am not familiar with SVI, but I will try that.

You have pointed out to reparameterize  `kappas`, but I am not sure about that!
Kappas and mus have a shape (N, 2) where N can be the number of row, sometime N=25000.

Can you write that part "reparameterize"  with code, also the last 4 lines written above ?

**Comment by fehiepsi** [comment]
@Nikolin4 Assuming that `locs` and `kappas` are loc and scale parameters of Normal distribution, then instead of
```
phi = numpyro.sample('phi', dist.Normal(mus[: ,0], kappas[: , 0])
```
, you can reparameterize as follows
```
phi = mus[:, 0] + kappas[:, 0] * numpyro.sample('phi_base', dist.Normal(0, 1).expand([N]))
```
But I don't know how to do a similar thing for `VonMises`. I think that you don't have to worry about it if `phi` and `psi` are observed. Is that the case here? If they are latent variables, then I don't think that MCMC will work at all because the dimensions of those latent variables are so large.

I guess they are observed variables, but your code does not say so. For example, in bnn example that you referred, `Y` is the observed node:
```
numpyro.sample("Y", dist.Normal(z3, sigma_obs), obs=Y)
```
I can see `Y` in your model input but I couldn't find it in the code. Could you provide more information? How does it relate to `phi` and `psi`?

**Comment by Nikolin4** [comment]
@fehiepsi I have updated the code above, and I can see that I didn't provide you with the right info.
Y contains both phi and psi.
We know that phi and psi are Von Misses distributed, and that's why I cannot use  `dist.Normal`.

**Comment by fehiepsi** [comment]
Hi @Nikolin4, thanks to your report, we have identified some issues of VonMises implementation. For follow-up questions such as how to make BNN work for a particular dataset, you can follow up in the [forum](https://forum.pyro.ai/), which will catch more eyes from Pyro community. We just use github issue for bug reports and feature requests.
